### PR TITLE
Limit the number of bulk values

### PR DESCRIPTION
**Pull request by c000**
Limit the number of bulk values to avoid the error.
> pq: got xxxxx parameters but PostgreSQL only supports 65535 parameters

### CODE REVIEW

**Comment by kamichidu** [comment]
tell me why you didn't fix about `DefaultCompiler` .
the same error will be occurred.

in my opinion,
each RDBMS (and its product version) has different placeholders limit allowed in a query.
when 1 patch has placeholders more than the limit, that error is not avoidable.
how about that introducing `MaxPatches` option for merging specific number of patches?

**Comment by c000** [comment]
Because I think it is impossible to avoid the error when using `Default Compiler`.
It might be theoretically possible, but I thought it was too complicated.

However, I agree that the number of placeholders varies depending on the RDBMS and their versions, I will change to add MaxPatches.

File: compiler.go [hunk 1]
```diff
@@ -140,9 +140,13 @@ func (*defaultCompiler) Compile(options *CompilerOptions) (sqlizers *SqlizerList
 	return sqlizers
 }
 
-type bulkCompiler struct{}
+type BulkCompilerWithOption struct {
```
Comment (kamichidu) [inline compiler.go]: i prefer `BulkCompilerWithOptions` like `CompilerOptions`  .

File: compiler.go [hunk 2]
```diff
@@ -10,7 +10,7 @@ import (
 var (
 	DefaultCompiler PatchCompiler = &defaultCompiler{}
 
-	BulkCompiler PatchCompiler = &bulkCompiler{}
+	BulkCompiler PatchCompiler = &BulkCompilerWithOption{MaxPatches: 1000}
```
Comment (kamichidu) [inline compiler.go]: default value should `0` (unlimited) .
the placeholders limit is 999 when use sqlite3 driver.

if we have to use this option should create new instance with flavored options.

File: compiler_test.go [hunk 3]
```diff
@@ -455,4 +455,136 @@ func TestBulkCompiler(t *testing.T) {
...
+						"testing",
+						[]string{"id", "name"},
+						[]interface{}{2, "b"},
+					),
+					InsertPatch(
+						"testing",
+						[]string{"id", "name"},
+						[]interface{}{3, "c"},
+					),
+				},
```
Comment (kamichidu) [inline compiler_test.go]: should test with 5 patches.

File: compiler_test.go [hunk 4]
```diff
@@ -455,4 +455,136 @@ func TestBulkCompiler(t *testing.T) {
...
+							"id": 2,
+						},
+					}),
+					UpdatePatch("testing", []string{"name"}, []interface{}{"a"}, &MapRowKey{
+						Table: "testing",
+						Key: map[string]interface{}{
+							"id": 3,
+						},
+					}),
+				},
```
Comment (kamichidu) [inline compiler_test.go]: 5 patches, please.

File: compiler_test.go [hunk 5]
```diff
@@ -455,4 +455,136 @@ func TestBulkCompiler(t *testing.T) {
...
+							"id": 2,
+						},
+					}),
+					DeletePatch("testing", &MapRowKey{
+						Table: "testing",
+						Key: map[string]interface{}{
+							"id": 3,
+						},
+					}),
+				},
```
Comment (kamichidu) [inline compiler_test.go]: 5 patches, please.

File: compiler.go (hunk 1 above)
Comment (c000) [reply compiler.go]: fixed

File: compiler.go (hunk 2 above)
Comment (c000) [reply compiler.go]: fixed

File: compiler_test.go (hunk 3 above)
Comment (c000) [reply compiler_test.go]: fixed

File: compiler_test.go (hunk 4 above)
Comment (c000) [reply compiler_test.go]: fixed

File: compiler_test.go (hunk 5 above)
Comment (c000) [reply compiler_test.go]: fixed

**Pull request merged**